### PR TITLE
Fix stale location manager entry on dynamic deletions

### LINKED ISSUE #3660: issues with dynamic insertion and deletion of array chare elements
Hello,

I am developing adaptive mesh refinement for [spectre](https://github.com/sxs-collaboration/spectre) (which uses charm++) and have encountered several issues when inserting new elements and destroying old elements of an array chare.

The issues could be caused by my missing or not understanding the documentation regarding dynamic array insertion/deletion such as calls to `beginInserting` and `doneInserting`.

I have put together several charm++ only examples that illustrate two issues I have encountered, as well as a complete 1D example of the AMR algorithm that I am trying to implement.

The first issue is reproduced in `dynamic_array_expansion`.  The issue is that some newly created elements execute all entry methods called from the beginning of execution instead of those called after their creation.  This example fails on the `main` branch but not on `v7.0.0`.  git bisection points to commit 5e26e07023 (which I note is also blamed for #3643).

The second issue is reproduced in `dynamic_array_hang`.  The issue is that the code hangs in some instances when an element has been deleted at some point and then recreated later on a different PE.  This fails on v7.0.0, but the first issue prevents this example from being tested on the main branch.

Please let me know if there is any additional information you need.

### PR DESCRIPTION
**Pull request by adityapb**
Adding tests from the spectre group #3662 

Resolves #3660 - The issue was caused by the location manager entry not being deleted correctly on dynamic deletions

### CODE REVIEW

**Review by ericjbohm — APPROVED** [review]

File: tests/charm++/dynamic_array_hang/dynamic_array_hang.C [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+//  - Test hangs on more than one PE for charm v7.0.0
+//
+//  - Test fails on latest because of the dynamic_insertion error before getting
+//    to the point v7.0.0 hangs
+//
+//  - It appears that when one of the initial elements is destroyed and then
+//    recreated in a later iteration on a PE other than PE 0, it never receives
+//    the ping.  (Could this be related to the load balancing issue (3605) that
+//    was addressed by charm++ PR 3614?)
+//
```
Comment (rbuch) [inline tests/charm++/dynamic_array_hang/dynamic_array_hang.C:54]: Same as earlier, should be removed if fixed in this PR.

File: tests/charm++/dynamic_array_expansion/dynamical_insertion.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Comment (rbuch) [inline tests/charm++/dynamic_array_expansion/dynamical_insertion.h:1]: `#pragma once` is non-standard, it should ideally be replaced with a more portable solution,

File: tests/charm++/random_1d_amr_example/random_amr.C [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+//      - If it neither want to join nor split, the element updates its data
+//        (i.e. determines its new neighbors).
+//
+// Current behaviour (Oct 25, 2022):
+//
+//  - hangs (or randomly fails) on more than two PEs for charm v7.0.0
+//    because of dynamic_array_hang
+//
+//  - fails on latest because of the dynamic_insertion error before getting
+//    to the point v7.0.0 hangs
```
Comment (rbuch) [inline tests/charm++/random_1d_amr_example/random_amr.C:78]: Same as other blocks.

File: tests/charm++/dynamic_array_expansion/dynamical_insertion.ci [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
```
Comment (rbuch) [inline tests/charm++/dynamic_array_expansion/dynamical_insertion.ci:2]: Presumably we want these tests to fall under the normal Charm++ license, so we should probably remove these blocks from all the files. The original authors of the tests have already implicitly agreed to this, but it's perhaps worth double checking with them that they're willing to contribute this under our usual license.

**Review by rbuch — DISMISSED** [review]
Mostly looks good, just a few minor quibbles. Overall, I think I'd prefer if this was rebased into two commits, one for the  change to `cklocation.C` and one that adds the tests.

**Comment by adityapb** [comment]
@rbuch - I think I have addressed all the comments. Let me know if you'd suggest any other changes

**Review by stwhite91 — COMMENTED** [review]
Can you put the issue # in the commit msg header, ie preface it with "Bug #3660: "?

Also I think we can name the tests more descriptively and consistently with the rest of the tests. Here are my suggestions:

dynamic_array_expansion  ->  dynamic insertion
dynamic_array_hang  ->  dynamic_insertion_deletion
random_1d_amr_example  -> amr_1d_random
simple_1d_amr  ->  amr_1d_simple

I would also suggest renaming the source files for each test similarly. Otherwise this all looks good to me

**Review by stwhite91 — APPROVED** [review]

**Pull request merged**